File: frontend/src/apis/mcpProviders.ts
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

import apiClient from './client'

export interface MCPProvider {
  key: string
  name: string
  name_en?: string
  description: string
  discover_url: string
  api_key_url: string
  token_field_name: string
  has_token: boolean
}

export interface MCPServer {
  id: string
  name: string
  description?: string
  type: string
  base_url?: string
  command?: string
  args?: string[]
  env?: Record<string, string>
  headers?: Record<string, string>
  is_active: boolean
  provider: string
  provider_url?: string
  logo_url?: string
  tags?: string[]
}

export interface MCPProviderKeysRequest {
  bailian?: string
  modelscope?: string
  mcp_router?: string
}

export interface MCPProviderKeysResponse {
  success: boolean
  message: string
}

export interface MCPProviderListResponse {
  providers: MCPProvider[]
}

export interface MCPServerListResponse {
  success: boolean
  message: string
  servers: MCPServer[]
  error_details?: string
}

export const mcpProviderApis = {
  /** Get list of MCP providers */
  getProviders: async (): Promise<MCPProviderListResponse> => {
    return apiClient.get<MCPProviderListResponse>('/mcp-providers')
  },

  /** Sync MCP servers from a provider */
  syncServers: async (providerKey: string): Promise<MCPServerListResponse> => {
    return apiClient.post<MCPServerListResponse>(`/mcp-providers/${providerKey}/servers`)
  },

  /** Update MCP provider API keys */
  updateKeys: async (data: MCPProviderKeysRequest): Promise<MCPProviderKeysResponse> => {
    return apiClient.put<MCPProviderKeysResponse>('/mcp-providers/keys', data)
  },
}

File: backend/app/schemas/mcp_providers.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class MCPServer(BaseModel):
    """MCP Server definition"""

    id: str  # Provider-prefixed ID: @bailian/{id}
    name: str  # Display name
    description: Optional[str] = None
    type: str  # "streamable-http" | "sse"
    base_url: Optional[str] = None  # Server URL
    command: Optional[str] = None  # For stdio type
    args: Optional[List[str]] = None  # Command arguments
    env: Optional[Dict[str, str]] = None  # Environment variables
    headers: Optional[Dict[str, str]] = None  # HTTP headers (auth)
    is_active: bool = True
    provider: str  # Provider name
    provider_url: Optional[str] = None
    logo_url: Optional[str] = None
    tags: Optional[List[str]] = None


class MCPProviderInfo(BaseModel):
    """MCP Provider metadata"""

    key: str
    name: str
    name_en: Optional[str] = None
    description: str
    discover_url: str
    api_key_url: str
    token_field_name: str
    has_token: bool = False  # Whether user has configured token


class MCPProviderListResponse(BaseModel):
    """Response for listing providers"""

    providers: List[MCPProviderInfo]


class MCPServerListResponse(BaseModel):
    """Response for listing servers"""

    success: bool
    message: str
    servers: List[MCPServer]
    error_details: Optional[str] = None


class MCPProviderKeysRequest(BaseModel):
    """Request to update MCP provider API keys"""

    bailian: Optional[str] = None
    modelscope: Optional[str] = None
    mcp_router: Optional[str] = None


class MCPProviderKeysResponse(BaseModel):
    """Response for MCP provider keys update"""

    success: bool
    message: str
